[microvm] Collapses common code between microvm and hyperlight

LucasGdosR opened the pull request:
The I/O thread code can be exactly the same. For now, hyperlight doesn't use every channel that's created, so they're left with a leading underscore.

The memory thread is not exactly the same. A closure is used to encapsulate the different code.

LucasGdosR committed:
[microvm] E: Merge Code for MicroVM and Hyperlight

The I/O thread code can be exactly the same. For now, hyperlight doesn't
use every channel that's created, so they're left with a leading
underscore.

The memory thread is not exactly the same. A closure is used to
encapsulate the different code.

LucasGdosR commented:
These changes have been incorporated into https://github.com/nanvix/nanvix/pull/856. They can be merged independently or together.

LucasGdosR commented:
Superseded by https://github.com/nanvix/nanvix/pull/856. It also colapses orchestrating logic.

LucasGdosR committed:
[microvm] E: Extract Snapshot Protocol

This commit extracts the snapshot protocol to orchestrator.rs.

LucasGdosR commented:
All the code regarding the snapshot protocol is now part of a separate file that's reused between the microvm and hyperlight. It will also use callback functions to deal with different implementations between the microvm and hyperlight, particularly for creating snapshots and loading snapshots.

csegarragonz commented on src/microvm/src/memory.rs:
```diff
@@ -0,0 +1,77 @@
+///
+/// # Returns
+///
+/// A handle to the I/O thread.
```
A handle to the memory thread.

ppenna commented on src/microvm/src/orchestrator.rs:
```diff
@@ -0,0 +1,275 @@
+    Receiver,
+    Sender,
+};
+use std::{
```
@LucasGdosR please address this.

LucasGdosR commented:
Many thanks for the review, and sorry for being sloppy with the description of the memory thread and the imports. I have corrected those and squashed the bug fixing commits.

csegarragonz commented on src/microvm/src/memory.rs:
```diff
@@ -0,0 +1,77 @@
+// Copyright(c) The Maintainers of Nanvix.
+// Licensed under the MIT License.
+
```
I would personally name this module `memory_thread` as it is more explainatory than `memory`.

Can we also add a module level block comment explaining why we need a memory_thread and what is its purpose?

Something like:

```rust
//! Bla bla
```

LucasGdosR replied:
I agree with the module name proposal. I used `memory` because the previous module for the io thread was called `io`, but I think that's also inappropriate. I'd change them both to include `_thread`. I also agree a module level block comment would make it clearer for newcomers, and it would also be useful in the `io_thread` module and the (to be renamed) `orchestrator` module.

ppenna replied:
Introduced module-level documentation.

ppenna replied:
Renamed modules.

csegarragonz commented on src/microvm/src/memory.rs:
```diff
@@ -0,0 +1,77 @@
+///
+/// # Description
+///
+/// Spawns a new memory thread.
```
Same as above. What is a memory thread, why do we need one?

ppenna replied:
Improved documentation.

csegarragonz commented on src/microvm/src/orchestrator.rs:
```diff
@@ -0,0 +1,275 @@
+///
+/// # Description
+///
+/// States relating to snapshots functionality. Snapshots may be loaded at PreBoot, and created at Paused.
```
I think this is not particular to snapshots, but to the VMM itself?

LucasGdosR replied:
I agree. For example, these states could be used for live migrations as well.

ppenna replied:
Improved documentation.

csegarragonz commented on src/microvm/src/orchestrator.rs:
```diff
@@ -0,0 +1,275 @@
+// Copyright(c) The Maintainers of Nanvix.
+// Licensed under the MIT License.
+
```
Maybe nit-pick, but I don't think `orchestrator` is the right name for this module neither? The way I see it this module handles the state machine for the VMM, and the state machine transitions.

LucasGdosR replied:
It holds all control channels between the main thread and all other threads, and those control channels are used for orchestrating the other threads.

For example, the vCPU thread can only resume running the MicroVM once the memory thread is done writing to the kernel's register, otherwise there would be a race condition where the vCPU thread might resume and pause again before the memory thread writes to the kernel. The logic to avoid these race conditions would be in this module, using those control channels.

I think this module holds the main thread's functionality of communicating with the other threads, and the state machine transitions based on this communication.

ppenna replied:
Keeping orchestrator for now.

csegarragonz reviewed (commented):
Some comments regarding naming, happy to hear other's opinions.

ppenna commented:
@LucasGdosR and @csegarragonz I am taking over this PR to get it merged.

ppenna committed:
[microvm] E: Rename I/O Thread Module

ppenna committed:
[microvm] E: Rename Memory Thread Module

ppenna commented on src/microvm/src/vmm/hyperlight/mod.rs:
```diff
@@ -254,54 +277,27 @@ impl Vmm {
-            }
-        });
+        let memory_thread: JoinHandle<Result<(), anyhow::Error>> =
+            memory::spawn(memory_thread_rx, memory_thread_tx, add_credit);
```
This suggestion is invalid.

ppenna commented on src/microvm/src/memory_thread.rs:
```diff
@@ -0,0 +1,84 @@
+                Ok(mut msg) => {
+                    profiler::timestamp_message!(
+                        &mut msg.payload,
+                        mem::offset_of!(syscall::LinuxDaemonMessage, payload)
+                            + mem::offset_of!(syscall::unistd::message::ReadResponse, buffer)
```
The suggestion is wrong.

ppenna commented on src/microvm/src/vmm/hyperlight/mod.rs:
```diff
@@ -18,22 +12,47 @@ extern crate kvm_ioctls;
+        ControlCommandResponse,
+        IoThread,
+    },
+    memory,
```
Fixed.

ppenna commented on src/microvm/src/vmm/hyperlight/mod.rs:
```diff
@@ -18,22 +12,47 @@ extern crate kvm_ioctls;
+        ControlCommand,
+        ControlCommandResponse,
+        IoThread,
+    },
```
Fixed.

csegarragonz reviewed (approved):
LGTM! Thanks @LucasGdosR, this will help a lot!
